draft-thomson-http-replay.md

```diff
@@ -287,6 +289,37 @@ forwarding the request until the TLS handshake completes, or send a 4NN (Too
 Early) status code in response.  A gateway that is uncertain about whether an
 origin server supports the `Early-Data` header field SHOULD disable early data.
 
+## Consistent Handling of Early Data {#be-consistent}
+
+Consistent treatment of a request that arrives in - or partially in - early data
+is critical to avoiding inappropriate processing of replayed requests.  Requests
+that might be processed prior to the handshake completing MUST be processed the
+same way, even if the handshake is complete when processing occurs.
+
+A TLS implementation MUST ensure that it is possible to correctly identify early
+data.  This is necessary even if early data arrives at the same time as the TLS
+Finished message from the client.  If early data is misclassified, requests
+might erroneously appear to be safe to handle.  Even though the handshake might
+complete successfully on this connection, this is no guarantee that the same
+early data was not processed prior to handshake completion on another
+connection where the early data was replayed.
+
+A server MUST NOT process requests that arrive in early data, or requests marked
+with the `Early-Data` header field unless it determines that the processing is
+free from side-effects.  Otherwise, the server MUST consistently use the same
+approach: either using the 4NN status code, or delaying processing until the
+handshake completes.
+
+A server MUST use the 4NN status code if the request is marked with the
+Early-Data header field.  Presence of the header field indicates that early data
+was used on a previous hop.  Completion of the TLS handshake on that hop isn't
+visible to the server.
+
+Servers can sometimes initiate processing on requests that are only partially
+received.  These considerations apply to any processing that the server might
+apply.  Parsing header fields and request routing is likely to be safe from
```

Comment:
Probably be explicit; e.g., "Parsing header fields -- but not acting upon them -- and ..."